### ISSUE TITLE
Enhancement: [switch-exhaustiveness-check] Add ability to ignore default case

### ISSUE THREAD

**Issue by jakeleventhal** [typescript-eslint/typescript-eslint#8247]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md

### Description

https://github.com/typescript-eslint/typescript-eslint/issues/3616#issuecomment-1673619717

This comment was never implemented. The case that is not possible to detect is case 1 from the original thread

This should throw an error because all cases are not enumerated (despite there being a fallthrough).

### Fail

```TypeScript
declare const literal: "a" | "b";

switch (literal) {
  case "a": break;
  default: break;
}
```

### Pass

```TypeScript
declare const literal: "a" | "b";

// Valid
switch (literal) {
  case "a": break;
  case "b": break;
}

// Valid
switch (literal) {
  case "a": break;
  case "b": break;
  default: break;
}
```

### Additional Info

Why is this useful? Say you want to define explicit behavior for each enum value. I then add an enum value to my type. Having an error in this case will be an extra catch that will allow me to ensure that I remember to update the switch statements. Essentially I want to be able to always define behavior for known types.

**Comment by JoshuaKGoldberg** [comment]
I think this is reasonable provided the option links to the existing caveat mentioned in https://typescript-eslint.io/rules/switch-exhaustiveness-check/#allowdefaultcaseforexhaustiveswitch-caveats. Maybe, `requireDefaultForUnion` would be a good name to mirror the existing `requireDefaultForNonUnion` option?

cc @typescript-eslint/triage-team?

> Link to the rule's documentation
> 
> https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md

By the way, nit: https://typescript-eslint.io/rules/switch-exhaustiveness-check is the place to look for rule's documentation. That `.md` file directly says in its second block: _"🛑 This file is source code, not the primary documentation location! 🛑"_. 🙂

**Comment by arka1002** [comment]
Hi :wave:, I just had one question.

What would the rule do if `allowDefaultCaseForExhaustiveSwitch` is set to `false` & `requireDefaultForUnion`(the proposed option) is set to `true`.

For example,  see [this playground link](https://typescript-eslint.io/play#ts=5.4.2&fileType=.tsx&code=C4TwDgpgBAYg9nKBeKAiAhqqAfNAjVAbgFgAoAYzgDsBnYKAIXQC9nk1Mp0bYETSyNAO4BLYOQAWUABRNWASigBvKGShRy3aBlQAuZWvVQ8AJwjoA1v3UBfQ5praC%2BpYfWnzVw3dLqAJhAAZugArgA2wC5uUAD0MVAAdEneZD5kQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1w2HImjQA9tEgAaLdqzbskYfHj6OAEUQAzUfHwBhYckQAxQwCiQqLiUogAylw8vGqu8D5m2AC%2BiQC6WilJQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)(which shows the default clause as invalid) & the 2nd code snippet of "Pass" in the OP(which says the default clause should be valid!). 

What will happen then ?

**Comment by Josh-Cena** [comment]
Isn't this issue asking for an option that enforces "the default case can only match `never` if the matched value is a union"? How is `requireDefaultForUnion` helpful? I think `allowDefaultCaseMatchUnionMember` or something sounds better.
